File: freshman year/CSE12/hw1/hw1.java
/****************************************************************************

                                                        Jamie Zhang
                                                        CSE12, WI15
                                                        Jan 9, 2015
                                                        cs12xre
                                Assignment One

File Name:      hw1.java
Description:    This program tests functions to display output strings and 
                numbers. 
****************************************************************************/

/**
 * The hw1 class is a direct port of hw1.c to java.
 * As you already know in java when you pass literal strings like
 * <P>
 *   writeline("a literal string\n", stream);
 * <P>
 * in C is considered a char[], but in java it's automatically
 * converted and treated as a String object.  Therefore 
 * the function writeline accepts literal strings and 
 * String types.
 */

import java.io.*;      // System.out

/*
* Class Name: hw1
* Class Description: This program takes in numbers or strings and outputs
* them to a display.
*/
public class hw1 {
   private static final int COUNT = 8;	// number of hex digits to display
   private static final int DECIMAL = 10;	// to indicate base 10
   private static final int HEX = 16;		// to indicate base 16

   private static final char digits[] = 	// used for ASCII conversion
   new String("0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ").toCharArray();

   private static int hexCounter = 0; // counter for the number hex digits

   //used to create array in baseout  
   private static final int NUMARRAYDIGITS = 1000; 

   /**
   * Takes in a positive number and displays in a given base.
   *
   * @param    number   Numeric value to be displayed.
   * @param    base     Base used to display number.
   * @param    stream   Where to display, likely System.out or System.err.
   */
   private static void baseout (int number, int base, PrintStream stream) {
      int numCopy = number; //makes a copy of the number
      int arrayIndex = 0; //index of array, used to store and print the digits

      //creates a large, empty array
      char[] arrayDigits = new char[NUMARRAYDIGITS];
    
      //makes the first 8 ints in the array 0s
      for(int i = 0; i < COUNT; i++)
      {
        arrayDigits[i] = digits[0];
      } 

      while(numCopy != 0)
      {
        //finds the digit by modding, going from back to front and 
        //records that digit as a char
        arrayDigits[arrayIndex] = digits[numCopy % base];
        //divides by the base to get the number without the digit that was 
        //just found
        numCopy = numCopy / base;

        //goes to the next index in the array
        arrayIndex++;
      }

      //if arrayIndex was 0 (number is 0), print 0
      if(arrayIndex == 0)
      {
         fputc(digits[0], stream);
      }	

      //if base 16
      else if(base == HEX)
      {
         arrayIndex = COUNT - 1; //first digit is at this index
  
         while(arrayIndex >= 0)
         {	    
            //finds and records the digit that needs to be printed next
            char digitToPrint = arrayDigits[arrayIndex];

            //calls on fputc to print the digit
            fputc(digitToPrint, stream);

            //decrements arrayIndex to print the next digit
            arrayIndex--;
         }
      }

      //any base except base 16
      else
      {
	 //first digit to be printed is at this index
         int lastDigitIndex = arrayIndex - 1;

         //while the number hasn't finished printing
         while(lastDigitIndex >= 0)
         {
            //finds and records the digit that needs to be printed next
            char digitToPrint = arrayDigits[lastDigitIndex];
 
            //calls on fputc to print the digit
            fputc(digitToPrint, stream);

            //decrements lastDigitIndex to print the next digit
            lastDigitIndex--;
         }
      }	
   }


   /**
   * Takes in a positive number and displays it in decimal.
   *
   * @param    number   positive numeric value to be displayed.
   * @param    stream   Where to display, likely System.out or System.err.
   */
   public static void decout (int number, PrintStream stream) {
      //calls baseout to print a number in base 10
      baseout(number, DECIMAL, stream);   
   }


   /**
   * Displays a single character.
   *
   * @param    CharToDisp  Character to display.
   * @param    stream      Where to display, likely System.out or System.err.
   */
   public static void fputc(char CharToDisp, PrintStream stream) {
      // Print a single character.
      stream.print (CharToDisp);   

      // Flush the system.out buffer, now. 
      stream.flush ();
   }

   /**
   * Takes in a positive number and displays it in hex.
   *
   * @param    number   A positive numeric value to be displayed in hex.
   * @param    stream   Where to display, likely System.out or System.err.
   */
   public static void hexout (int number, PrintStream stream) {
      // Output "0x" for hexidecimal.
      writeline ("0x", stream);
     
      //calls baseout with base 16
      baseout (number, HEX, stream);
   }


   /**
   * Prints out a newline character.
   * @param    stream  Where to display, likely System.out or System.err.
   *
   */
   public static void newline ( PrintStream stream ) {
      //outputs a new line character  
      writeline("\n", stream);
   }


   /**
   * Prints out a string.
   *
   * @param    message  A string to print out.
   * @param    stream   Where to display, likely System.out or System.err.
   * @return   <code>int</code> - The length of the string.
   */
   public static int writeline (String message, PrintStream stream) {        
      int index = 0; //what character to print next

      //prints the next character if there are still characters left to 
      //print
      while(index < message.length())
      {
         fputc(message.charAt(index), stream); //prints chars
         index++;
      }

      return message.length(); //returns length of message
   }


  public static void main( String[] args ) {
    System.err.print("Zdravei Sviat\n");

    writeline ("Hello World", System.out);
    newline(System.out);

    decout (123, System.out);
    newline(System.out);

    decout (0, System.out);
    newline(System.out);

    hexout (0xFEEDDAD, System.out);
    newline(System.out);
  }
}

File: freshman year/CSE12/hw1/hw1.c
/****************************************************************************

                                                        Jamie Zhang
                                                        CSE12, WI15
                                                        Jan 9, 2015
                                                        cs12xre
                                Assignment One

File Name:      hw1.c
Description:    This program tests functions to display output strings and 
                numbers. 
****************************************************************************/
#include <stdio.h>

#define COUNT 8		/* number of hex digits to display */
#define DECIMAL 10	/* to indicate base 10 */
#define HEX 16		/* to indicate base 16 */

/* array used for ASCII conversion */
const char digits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";


/*--------------------------------------------------------------------------
Function Name:         writeline
Purpose:               Used to output characters to a display.
Description:           Uses a loop to print a character at a time until
                       all the characters are printed.
Input:                 message - what to be displayed
                       stream - where to display (stdout, stderr, etc)
Result:                A message is displayed.
                       int - the length of the message
--------------------------------------------------------------------------*/
int writeline (const char * message, FILE * stream) {
   int index = 0; /* counter used when printing characters */

   /* prints the next character if there are still characters left to 
      print */
   while(message[index]) {
      fputc(message[index], stdout); /* prints chars */
      index++;
   }

   return index; /* returns length of message */
}


/*--------------------------------------------------------------------------
Function Name:         baseout
Purpose:               Takes a number and prints it in a specified base.
Description:           Separates the number into its different digits in the 
                       base, converts the digits from int to char, and puts
		       them into a char array; then prints the digits one
		       at a time to display the number.
Input:                 number - a number to be displayed
                       base - the base the number should be displayed in
		       stream - where to display
Result:                A number in base base is displayed.
                       void - no return value
--------------------------------------------------------------------------*/
void baseout (int number, int base, FILE * stream) {
   int numCopy = number; /* makes a copy of the number */
   int arrayIndex = 0; /* index of array, used to store and print the digits */

   /* creates a large, empty array */
   char arrayDigits[BUFSIZ];

   /* makes the first 8 ints in the array 0s */   
   int i;
   for(i = 0; i < COUNT; i++)
   {
      arrayDigits[i] = digits[0];
   } 

   /* stores digits */
   while(numCopy != 0)
   {
      /* finds the digit by modding, going from back to front and 
         records that digit as a char */      
      arrayDigits[arrayIndex] = digits[numCopy % base];
      /* divides by the base to get the number without the digit that was 
         just found */
      numCopy = numCopy / base;

      /* goes to the next index in the array */
      arrayIndex++;
   }

   /* if arrayIndex was 0 (number is 0), print 0 */
   if(arrayIndex == 0)
   {
       fputc(digits[0], stream);
   }	

   /* if base 16 */
   else if(base == HEX)
   {
       arrayIndex = COUNT - 1; /* first digit is at this index */
  
       while(arrayIndex >= 0)
       {	    
          /* finds and records the digit that needs to be printed next */
          char digitToPrint = arrayDigits[arrayIndex];

          /* calls on fputc to print the digit */
          fputc(digitToPrint, stream);

          /* decrements arrayIndex to print the next digit */
          arrayIndex--;
       }
   }  

   /* any other base */
   else
   {
      /* first digit to be printed is at this index */	
      int lastDigitIndex = arrayIndex - 1;

      /* while the number hasn't finished printing */
      while(lastDigitIndex >= 0)
      {
         /* finds and records the digit that needs to be printed next */
         char digitToPrint = arrayDigits[lastDigitIndex];
 
         /* calls on fputc to print the digit */
         fputc(digitToPrint, stream);

         /* decrements lastDigitIndex to print the next digit */
         lastDigitIndex--;
      }
   }
}


/*--------------------------------------------------------------------------
Function Name:         decout
Purpose:               This function is used to print a number in base 10. 
                       The number is displayed in the parameter stream.
Description:           Calls the baseout function with base 10.
Input:                 number - the number to be displayed
                       stream - where to display the number 
		                (stdout, stderr, etc)
Result:                A number in base 10 is displayed.
                       void - no return value
--------------------------------------------------------------------------*/
void decout (unsigned int number, FILE * stream) {
   /* calls baseout to print a number in base 10 */
   baseout(number, DECIMAL, stream);
}


/*--------------------------------------------------------------------------
Function Name:         hexout
Purpose:               Prints a number in base 16 to the parameter FILE stream
Description:           Goal is achieved via delegating to the baseout function
Input:                 number:  the number to display
                       stream:  where to display, likely stdout or stderr
Result:                Number in base 16 is displayed.
                       No return value.
--------------------------------------------------------------------------*/
void hexout (unsigned int number, FILE * stream) {

   /* Output "0x" for hexidecimal. */
   writeline ("0x", stream);

   /* calls baseout to print a number in base 16 */
   baseout (number, HEX, stream);
}


/*--------------------------------------------------------------------------
Function Name:         newline
Purpose:               Used to print a new line to the parameter stream.
Description:           Calls writeline to display a new line character in the 
                       parameter stream.
Input:                 stream - where to display (stdout, stderr, etc)
Result:                A new line character is displayed.
                       void - no return value
--------------------------------------------------------------------------*/
void newline (FILE * stream) {
   /* Prints a new line character. */
   writeline("\n", stream);
}


int main (int argc, char *const* argv) {
   writeline ("Hello World", stdout);

   fprintf (stderr, "Zdravei Sviat\n");
   newline(stdout);

   decout (123, stdout);
   newline(stdout);

   decout (0, stdout);
   newline(stdout);

   hexout (0xFEEDDAD, stdout);
   newline(stdout);

   return 0;
}
